src/Education/XamlGenerator/IRuleHelpXamlBuilder.cs

```diff
@@ -0,0 +1,472 @@
+﻿/*
+ * SonarLint for Visual Studio
+ * Copyright (C) 2016-2023 SonarSource SA
+ * mailto:info AT sonarsource DOT com
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+
+using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.IO;
+using System.Text;
+using System.Windows.Documents;
+using System.Windows.Markup;
+using System.Xml;
+using SonarLint.VisualStudio.Rules;
+
+namespace SonarLint.VisualStudio.Education.XamlGenerator
+{
+    public interface IRuleHelpXamlBuilder
+    {
+        /// <summary>
+        /// Generates a XAML document containing the help information for the specified rule
+        /// </summary>
+        /// <remarks>Assumes that the <see cref="IRuleHelp.HtmlDescription"/> is parseable as XML</remarks>
+        FlowDocument Create(IRuleHelp ruleHelp);
+    }
+
+    internal partial class RuleHelpXamlBuilder : IRuleHelpXamlBuilder
+    {
+        private const string XamlNamespace = "http://schemas.microsoft.com/winfx/2006/xaml/presentation";
+
+        private XmlWriter writer;
+        private XmlReader reader;
+
+        /// <summary>
+        /// Stack of currently open XAML elements
+        /// </summary>
+        /// <remarks>We need some information about the current structure so we can check whether some 
+        /// operations are valid e.g. can be we add text to the current element?</remarks>
+        private Stack<XamlOutputElementInfo> outputXamlElementStack = new Stack<XamlOutputElementInfo>();
+
+        /// <summary>
+        /// Used to add background colour to alternate rows in a table
+        /// </summary>
+        private bool tableAlternateRow;
+
+        public FlowDocument Create(IRuleHelp ruleHelp)
+        {
+            var xaml = CreateXamlString(ruleHelp);
+
+            using (var stringReader = new StringReader(xaml))
+            {
+                var xamlReader = XmlReader.Create(stringReader);
+                var flowDocument = (FlowDocument)XamlReader.Load(xamlReader);
+                return flowDocument;
+            }
+        }
+
+        internal /* for testing */ string CreateXamlString(IRuleHelp ruleHelp)
```

Comment:
Added to make testing the produced XAML easier (it's easier to check a string than the XAML object graph). However, I've stashed the few low-level tests that I wrote, so this method isn't used by the current test.